### ISSUE TITLE
AWS PrivateLink now supports Private DNS names for internal and 3rd party services

### ISSUE THREAD

**Issue by ewbankkit** [hashicorp/terraform-provider-aws#11510]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description

You can now access AWS PrivateLink based services privately from within your VPC using [Private DNS names](https://docs.aws.amazon.com/vpc/latest/userguide/vpce-interface.html#vpce-private-dns).

### New or Affected Resource(s)

* [`aws_vpc_endpoint`](https://www.terraform.io/docs/providers/aws/r/vpc_endpoint.html)
* [`aws_vpc_endpoint_service`](https://www.terraform.io/docs/providers/aws/r/vpc_endpoint_service.html)

### Potential Terraform Configuration

```hcl
# Copy-paste your Terraform configurations here - for large Terraform configs,
# please use a service like Dropbox and share a link to the ZIP file. For
# security, you can also encrypt the files using our GPG public key.
```

### References

[Announcement](https://aws.amazon.com/about-aws/whats-new/2020/01/aws-privatelink-supports-private-dns-names-internal-3rd-party-services/).
[User guide](https://docs.aws.amazon.com/vpc/latest/userguide/vpce-interface.html#vpce-private-dns).

Requires[AWS SDK v1.27.1](https://github.com/aws/aws-sdk-go/releases/tag/v1.27.1):
* https://github.com/terraform-providers/terraform-provider-aws/pull/11501

**Comment by isen-ng** [comment]
This feature would be a great boon for implementing SSL over private links.

**Comment by mattsawyer77** [comment]
@ewbankkit I think the affected resource is actually [aws_vpc_endpoint_service](https://www.terraform.io/docs/providers/aws/r/vpc_endpoint_service.html), no? The aws_vpc_endpoint already has support for private DNS.

**Comment by ewbankkit** [comment]
@mattsawyer77 Correct, there's a flag on the VPC Endpoint that determines whether or not Private DNS is enabled but the actual DNS name is specified on the VPC Endpoint Service.
I've updated above.
Thanks.

**Comment by craigedmunds** [comment]
Have raised a PR for this as it's affecting us too!

**Comment by craigedmunds** [comment]
How do we get a review for this PR?

**Comment by bondezbond** [comment]
is there anyway to get this reviewed and merged in the next release?

**Comment by patcable** [comment]
Want to 👍 this as well - though, it'd be good to get the attributes from the DNS entry so I can use them with another provider to actually set the DNS record.  I did a bit of digging and i wasn't sure what the "right way" to expose those variables were - have done a few things w/ AWS API in golang, but not with terraform. It looks like the `ServiceConfiguration` struct references `PrivateDnsNameConfiguration` that has those values. Could I make some variables like `private_dns_name_record_{name, type, value}` that map to `*ec2.PrivateDnsNameConfiguration.{Name,Type,Value}` to get that change in?

**Comment by bflad** [comment]
Support for configurable `private_dns_name` and adding a `private_dns_name_configuration` attribute in the `aws_vpc_endpoint_service` resource has been merged and will release with version 3.23.0 of the Terraform AWS Provider. Thanks to @craigedmunds and @stijndehaes for their efforts here. 👍
